src/go/networking/pkg/portproxy/server_test.go

```diff
@@ -24,18 +24,101 @@ import (
 	"net/http"
 	"syscall"
 	"testing"
+	"time"
 
 	"github.com/docker/go-connections/nat"
 	"github.com/rancher-sandbox/rancher-desktop/src/go/guestagent/pkg/types"
 	"github.com/rancher-sandbox/rancher-desktop/src/go/networking/pkg/portproxy"
-	"github.com/sirupsen/logrus"
 	"github.com/stretchr/testify/require"
 	"golang.org/x/net/nettest"
 )
 
-func TestNewPortProxy(t *testing.T) {
-	logrus.SetLevel(logrus.DebugLevel)
+func TestNewPortProxyUDP(t *testing.T) {
+	testServerIP, err := availableIP()
+	require.NoError(t, err, "cannot continue with the test since there are no available IP addresses")
+
+	remoteAddr := net.JoinHostPort(testServerIP, "0")
+	targetAddr, err := net.ResolveUDPAddr("udp", remoteAddr)
+	require.NoError(t, err)
+	targetConn, err := net.ListenUDP("udp", targetAddr)
+	require.NoError(t, err)
+
+	t.Logf("created the following UDP target listener: %s", targetConn.LocalAddr().String())
+
+	localListener, err := nettest.NewLocalListener("unix")
+	require.NoError(t, err)
+	defer localListener.Close()
+
+	proxyConfig := &portproxy.ProxyConfig{
+		UpstreamAddress: testServerIP,
+		UDPBufferSize:   1024,
+	}
+	portProxy := portproxy.NewPortProxy(localListener, proxyConfig)
+	go portProxy.Start()
+
+	_, testPort, err := net.SplitHostPort(targetConn.LocalAddr().String())
+	require.NoError(t, err)
+
+	port, err := nat.NewPort("udp", testPort)
+	require.NoError(t, err)
+
+	portMapping := types.PortMapping{
+		Remove: false,
+		Ports: nat.PortMap{
+			port: []nat.PortBinding{
+				{
+					HostIP:   "127.0.0.1",
+					HostPort: testPort,
+				},
+			},
+		},
+	}
+	t.Logf("sending the following portMapping to portProxy: %+v", portMapping)
+	err = marshalAndSend(localListener, portMapping)
+	require.NoError(t, err)
+
+	// indicate when UDP mappings are ready
+	ready := make(chan struct{})
+	var udpMappings map[int]*net.UDPConn
+	go func() {
+		for {
+			udpMappings = portProxy.UDPPortMappings()
+			if len(udpMappings) > 0 {
+				close(ready)
```

Comment:
You are correct, my original approach was polling for Port (before implementing `UDPPortMappings`) was doing things in a goroutine, but since we have the function, there is no need for a goroutine, removed.